external/fv3fit/fv3fit/pytorch/predict.py

```diff
@@ -1,18 +1,54 @@
+from fv3fit._shared.predictor import Dumpable, Loadable
 from .._shared.scaler import StandardScaler
 import numpy as np
 import torch
 import torch.nn as nn
 import xarray as xr
-from fv3fit._shared import Dumpable
-from typing import Any, Dict, Hashable, Iterable, Mapping, Tuple
+from typing import Hashable, Iterable, Mapping, Tuple, TypeVar, Type, IO, Protocol
+import zipfile
 from fv3fit.pytorch.system import DEVICE
+import os
+import yaml
+import vcm
 
 
-class PytorchModel(Dumpable):
+L = TypeVar("L", bound="BinaryLoadable")
+
+
+class BinaryLoadable(Protocol):
```

Comment:
Why is Loadable an ABC  and BinaryLoadable a protocol?

Personally I'd prefer the latter for both so that we don't need to inherit from it explicitly.

Reply:
Quoting from https://peps.python.org/pep-0544/

> The same problem appears with user-defined ABCs: they must be explicitly subclassed or registered. This is particularly difficult to do with library types as the type objects may be hidden deep in the implementation of the library. Also, extensive use of ABCs might impose additional runtime costs.

> The intention of this PEP is to solve all these problems by allowing users to write the above code without explicit base classes in the class definition, allowing Bucket to be implicitly considered a subtype of both Sized and Iterable[int] by static type checkers using structural [[wiki-structural]](https://peps.python.org/pep-0544/#wiki-structural) subtyping:

Reply:
Because `Dumpable` and `Loadable` are part of the definition of `Predictor`, which we've always had as an ABC for a little extra safety in ensuring we implement the required methods when they're made. The drawback isn't much of one because of the level of integration we already need for code that produces these classes (e.g. registering the functions which produce them).

On the other hand, `BinaryLoadable` is only written to serve as part of the type hint for `load_mapping`.